### Regulare Expressions: 

* It's a Matching Techniques. 
* It can extract the pattern from large documents easily! 
* Like Emails, documents and more......

Follow this: https://www.python-engineer.com/posts/regular-expressions/

In [2]:
import re

#### Regular Expression is Case Senstitive 💔

In [4]:
test_string = '123abc456789abc123ABC'

""" If you want to retrive abc from the test_string """ 

pattern = re.compile(r'abc') 

matches = pattern.finditer(test_string) 

for match in matches: 
    print(match)

<re.Match object; span=(3, 6), match='abc'>
<re.Match object; span=(12, 15), match='abc'>


In [6]:
# alternate method to run the exat code!

matches = re.finditer(r'abc', test_string) 
for match in matches: 
    print(match)

<re.Match object; span=(3, 6), match='abc'>
<re.Match object; span=(12, 15), match='abc'>


#### r - Raw String. 

Look at the bellow example

In [9]:
a = '\t aravind' 
b = r'\t aravind'
print(a)
print(b, 'It print whatever the word that present inside the quotes')

	 aravind
\t aravind It print whatever the word that present inside the quotes


## 1. Methods to Search for Matches

some matching methods are 

* match(): Determine if the RE matches at the beginning of the string.
* search(): Scan through a string, looking for any location where this RE matches.
* findall(): Find all substrings where the RE matches, and returns them as a list.
* finditer(): Find all substrings where the RE matches, and returns them as an iterator.

In [11]:
test_string = '123abc456789abc123ABC'

""" Findall returns the string, and it will search everywhere """ 

pattern = re.compile(r'abc') 

matches = pattern.findall(test_string) 

a = [ i for i in matches ] 
a

['abc', 'abc']

In [15]:
test_string = '123abc456789abc123ABC'

""" Match returns the string, that will apperar in the beginning of the string """ 

pattern = re.compile(r'abc') 

matches = pattern.match(test_string) 

print(matches) 

""" It gives none, because match look only begining of the strings """

None


' It gives none, because match look only begining of the strings '

In [19]:
test_string = '123abc456789abc123ABC'

""" search returns the firstfinding string, and it will search everywhere """ 

pattern = re.compile(r'abc') 

matches = pattern.search(test_string) 

matches

<re.Match object; span=(3, 6), match='abc'>

In [31]:
test_string = '123abc456789abc123ABC'

""" Finditer returns the string for iterator, and it will search everywhere """ 

pattern = re.compile(r'abc') 

matches = pattern.finditer(test_string) 

for match in matches: 
    print(match)

""" And after using find iter() we can do certain operations like group, start, end, span """

<re.Match object; span=(3, 6), match='abc'>
<re.Match object; span=(12, 15), match='abc'>


' And after using find iter() we can do certain operations like group, start, end, span '

## 2. Methods on a Finditer match

In [36]:
# match.span(), we will get the tuple here 
test_string = '123abc456789abc123ABC'

""" Finditer returns the string for iterator, and it will search everywhere """ 

pattern = re.compile(r'abc') 

matches = pattern.finditer(test_string) 
for match in matches: 
    print(match.span())

    

(3, 6)
(12, 15)


In [38]:
# finditer()
my_string = 'abc123ABC123abc'
pattern = re.compile(r'123')
matches = pattern.finditer(my_string)
for match in matches:
    print(match)
    print(match.span(), match.start(), match.end()) # start and stop gives the index numbes
    print(match.group()) # It just group all 

<re.Match object; span=(3, 6), match='123'>
(3, 6) 3 6
123
<re.Match object; span=(9, 12), match='123'>
(9, 12) 9 12
123


## 3. Meta Characters

* Metacharacters are characters with a special meaning:
* All meta characters: **. ^ $ * + ? { } [ ] \ | ( )**

* . - We want to look for any character (except a new line character) 
* ^ - Starts with '^hello' 
* '$' - Ends with Words'$' (dollar Symbol look the metadata)
* '*' - '
* '+' - One or more occurence. 
* {} - Exactly specified number of characters. 
* [] - A set of characters [a-z] 
* | - Either or 'falls or stays' 
* () - Capture and group 
* \ - special sequence or escape characters '\d'
*** 
### More Metacharacters / Special Sequences
A special sequence is a \ followed by one of the characters in the list below, and has a special meaning:

* \d :Matches any decimal digit; this is equivalent to the class [0-9].
* \D : Matches any non-digit character; this is equivalent to the class [^0-9].
* \s : Matches any whitespace character;
* \S : Matches any non-whitespace character;
* \w : Matches any alphanumeric (word) character; this is equivalent to the class [a-zA-Z0-9_].
* \W : Matches any non-alphanumeric character; this is equivalent to the class [^a-zA-Z0-9_].
* \b Returns a match where the specified characters are at the beginning or at the end of a word r"\bain" r"ain\b"
* \B Returns a match where the specified characters are present, but NOT at the beginning (or at the end) of a word r"\Bain" r"ain\B"
* \A Returns a match if the specified characters are at the beginning of the string "\AThe"
* \Z Returns a match if the specified characters are at the end of the string "Spain\Z"



In [45]:
# 1 '.' - Dot print very characther excpet new line 

test_string = '123abc456789abc123ABC'

pattern = re.compile(r'.') 

matches = pattern.finditer(test_string) 

for i in matches: 
    print(i.group())

1
2
3
a
b
c
4
5
6
7
8
9
a
b
c
1
2
3
A
B
C


In [58]:
# \ this is the search characther

test_string = '123abc456789abc123ABC.'

pattern = re.compile(r'\.') 
matches = pattern.finditer(test_string) 
for i in matches: 
    print(i)

<re.Match object; span=(21, 22), match='.'>


In [51]:
# ^ this ensures the starting charcters 

test_string = '123abc456789abc123ABC.'
pattern = re.compile(r'^123') 
matches = pattern.finditer(test_string) 
for i in matches: 
    print(i)

<re.Match object; span=(0, 3), match='123'>


In [55]:
# $ this ensures the ending charcters 

test_string = '123abc456789abc123ABC'
pattern = re.compile(r'ABC$') 
matches = pattern.finditer(test_string) 
for i in matches: 
    print(i)

<re.Match object; span=(18, 21), match='ABC'>


## 4. More Special Characters

In [59]:
# \d any decimal digits

test_string = 'hello 123_ heyho hohey'
pattern = re.compile(r'\d')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)

<re.Match object; span=(6, 7), match='1'>
<re.Match object; span=(7, 8), match='2'>
<re.Match object; span=(8, 9), match='3'>


In [69]:
# \D any non decimal digits (except \d )

test_string = 'hello 123_ heyho hohey'
pattern = re.compile(r'\D')
matches = pattern.finditer(test_string)
# for match in matches:
#      #print(match)

In [67]:
 # \s any whitespace character 

test_string = 'hello 123_ heyho hohey'
pattern = re.compile(r'\s')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)
print('') 

# \S Non whitespace character

test_string = 'hello 123_ heyho hohey'
pattern = re.compile(r'\S')
matches = pattern.finditer(test_string)
# for match in matches:
#     print(match)

<re.Match object; span=(5, 6), match=' '>
<re.Match object; span=(10, 11), match=' '>
<re.Match object; span=(16, 17), match=' '>



In [70]:
 # \w  Matches any alphanumeric (word) character; 

test_string = 'hello 123_ heyho hohey'
pattern = re.compile(r'\w')
matches = pattern.finditer(test_string)
# for match in matches:
#     print(match)
print('') 

# \W  Matches any non alphanumeric (word) character; 

test_string = 'hello 123_ heyho hohey'
pattern = re.compile(r'\W')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)


<re.Match object; span=(5, 6), match=' '>
<re.Match object; span=(10, 11), match=' '>
<re.Match object; span=(16, 17), match=' '>


In [76]:
 # \b it will find the begining of the block and not only begining if the string having spaces after the space the words is matched it will print

test_string = 'hello 123_ heyho hohey!'
pattern = re.compile(r'\bhey')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)
print('') 

# \B oppsitie to that, it will print end of the block 

test_string = 'hello 123_ heyho hobro'
pattern = re.compile(r'\Bbro')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)

<re.Match object; span=(11, 14), match='hey'>

<re.Match object; span=(19, 22), match='bro'>


## 5. Sets

A set is a set of characters inside a pair of square brackets [] with a special meaning. Append multiple conditions back-to back, e.g. [aA-Z].
A ^ (caret) inside a set negates the expression.
A - (dash) in a set specifies a range if it is in between, otherwise the dash itself.

Examples:
- [arn] Returns a match where one of the specified characters (a, r, or n) are present
- [a-n] Returns a match for any lower case character, alphabetically between a and n
- [^arn] Returns a match for any character EXCEPT a, r, and n
- [0123] Returns a match where any of the specified digits (0, 1, 2, or 3) are present
- [0-9] Returns a match for any digit between 0 and 9
- 0-5 Returns a match for any two-digit numbers from 00 and 59
- [a-zA-Z] Returns a match for any character alphabetically between a and z, lower case OR upper case

In [92]:
# Matching multiple single character 

test_string = 'hello 123_AARV'

pattern = re.compile(r'[a-z0-9A-B]') # all the characters from a - z 
matches = pattern.finditer(test_string)
 
for i in matches: 
    print(i)

<re.Match object; span=(0, 1), match='h'>
<re.Match object; span=(1, 2), match='e'>
<re.Match object; span=(2, 3), match='l'>
<re.Match object; span=(3, 4), match='l'>
<re.Match object; span=(4, 5), match='o'>
<re.Match object; span=(6, 7), match='1'>
<re.Match object; span=(7, 8), match='2'>
<re.Match object; span=(8, 9), match='3'>
<re.Match object; span=(10, 11), match='A'>
<re.Match object; span=(11, 12), match='A'>


## 6. Quantifier: 

* '*' : 0 or more
* '+' : 1 or more
* ? : 0 or 1, used when a character can be optional
* {4} : exact number
* {4,6} : range numbers (min, max)

In [97]:
test_string = 'hello_123' 

pattern = re.compile(r'\d*') # \d matches any decimal digits  
matches = pattern.finditer(test_string) 
for i in matches: 
    print(i)

<re.Match object; span=(0, 0), match=''>
<re.Match object; span=(1, 1), match=''>
<re.Match object; span=(2, 2), match=''>
<re.Match object; span=(3, 3), match=''>
<re.Match object; span=(4, 4), match=''>
<re.Match object; span=(5, 5), match=''>
<re.Match object; span=(6, 9), match='123'>
<re.Match object; span=(9, 9), match=''>


In [101]:
test_string = 'hello123' 

pattern = re.compile(r'_?\d+') # \? optional and + 1 or more digit underscore it finds that optional
matches = pattern.finditer(test_string) 
for i in matches: 
    print(i)

<re.Match object; span=(5, 8), match='123'>


In [108]:
test_string = 'hello123' 

pattern = re.compile(r'\d{1,2}') 
matches = pattern.finditer(test_string) 
for i in matches: 
    print(i)

<re.Match object; span=(5, 7), match='12'>
<re.Match object; span=(7, 8), match='3'>


#### Small Pratice

In [115]:
dates = '''
01.04.2020

2020.04.01

2020-04-01
2020-05-23
2020-06-11
2020-07-11
2020-08-11

2020/04/02

2020_04_04
2020_04_04
'''

# Here all the dates are in different different annotations, we need to get them in a right mannner, let's do 

pattern = re.compile(r'\d\d\d\d.\d\d.\d\d') 
pattern = re.compile(r'\d\d\d\d-\d\d-\d\d')
pattern = re.compile(r'\d\d\d\d/\d\d/\d\d')
pattern = re.compile(r'\d\d.\d\d.\d\d\d\d')
matches = pattern.finditer(dates) 
for i in matches: 
    print(i) 

<re.Match object; span=(1, 11), match='01.04.2020'>


In [117]:
# we will use the set for multiple retrival also using sets

pattern = re.compile(r'\d\d\d\d[/-]\d\d[-/]\d\d') 
matches = pattern.finditer(dates)
for i in matches: 
    print(i)

<re.Match object; span=(25, 35), match='2020-04-01'>
<re.Match object; span=(36, 46), match='2020-05-23'>
<re.Match object; span=(47, 57), match='2020-06-11'>
<re.Match object; span=(58, 68), match='2020-07-11'>
<re.Match object; span=(69, 79), match='2020-08-11'>
<re.Match object; span=(81, 91), match='2020/04/02'>


In [119]:
# Now I am only looking for dates only in may and june 

pattern = re.compile(r'\d\d\d\d[-/]0[4-6][-/]\d\d')
matches = pattern.finditer(dates)
for i in matches: 
    print(i)

<re.Match object; span=(25, 35), match='2020-04-01'>
<re.Match object; span=(36, 46), match='2020-05-23'>
<re.Match object; span=(47, 57), match='2020-06-11'>
<re.Match object; span=(81, 91), match='2020/04/02'>


In [122]:
# let's minimize the code

pattern = re.compile(r'\d{4}[-/]0[5,7][-/]\d{2}') # Now we use quantifies. . 
matches = pattern.finditer(dates)
for i in matches: 
    print(i)

<re.Match object; span=(36, 46), match='2020-05-23'>
<re.Match object; span=(58, 68), match='2020-07-11'>


## 7. Conditions:

In [133]:
my_string = """
hello world
1223 
2020-05-20
Mr Simpson
Mrs Simpson
Mr. Brown
Ms Smith
Mr. T
"""
# We want to extract only names here :) 

pattern = re.compile(r'[a-zA-Z.]') # My code
pattern = re.compile(r'(Mr|Ms|Mrs)\.?\s\w+') # condition based codes 
matches = pattern.finditer(my_string) 

for i in matches: 
    print(i)

<re.Match object; span=(30, 40), match='Mr Simpson'>
<re.Match object; span=(41, 52), match='Mrs Simpson'>
<re.Match object; span=(53, 62), match='Mr. Brown'>
<re.Match object; span=(63, 71), match='Ms Smith'>
<re.Match object; span=(72, 77), match='Mr. T'>


## 8. Grouping

In [154]:
emails = """
hello world
1223 
2020-05-20
Mr Simpson
Mrs Simpson
Mr. Brown
Ms Smith
Mr. T
pythonengineer@gmail.com
Python-engineer@gmx.de
python-engineer123@my-domain.org
"""
# I only want to extract the emails from the strings


pattern = re.compile(r'[a-zA-Z0-9-]+@[a-zA-Z-]+\.[a-zA-Z]+') # + 1 or more and it combines everything in a line 
pattern = re.compile(r'[a-zA-Z0-9-]+@[a-zA-Z-.]+')
pattern = re.compile(r'([a-zA-Z0-9-]+)@([a-zA-Z-]+)\.([a-zA-Z]+)') # we grouped over parenthesis 
matches = pattern.finditer(emails) 

for i in matches: 
    print(i.group(0))
    # print(i.group(1))
    # print(i.group(2))
    # print(i.group(3))

pythonengineer@gmail.com
Python-engineer@gmx.de
python-engineer123@my-domain.org


## 9. Modifications: 

There are two methods to modify the code 
* split(): Split the string into a list, splitting it wherever the RE matches
* sub(): Find all substrings where the RE matches, and replace them with a different string

In [159]:
# split
my_string = '123abc456789abc123ABC'

pattern = re.compile(r'abc')
matches = pattern.finditer(my_string) 

splitted = pattern.split(my_string) 
print(splitted)



['123', '456789', '123ABC']


In [161]:
# sub

test_string = 'hello world, you are the best wolrd' 

pattern = re.compile(r'world') 
subbed_string = pattern.sub('planet', test_string) 

print(subbed_string) # It just look the sentences and modifies into given word .... 

hello planet, you are the best wolrd


#### Small Example 2

In [173]:
urls = """
http://python-engineer.com
https://www.python-engineer.org
http://www.pyeng.net
"""


pattern = re.compile(r'https?://(www\.)?([a-zA-Z-]+)(\.[a-zA-Z]+)')
matches = pattern.finditer(urls) 

for match in matches: 
    print(match)
    
subbed_urls = pattern.sub(r'\2\3', urls) # we just called group 2 and group 3 
print(subbed_urls)

<re.Match object; span=(1, 27), match='http://python-engineer.com'>
<re.Match object; span=(28, 59), match='https://www.python-engineer.org'>
<re.Match object; span=(60, 80), match='http://www.pyeng.net'>

python-engineer.com
python-engineer.org
pyeng.net



## 10. Compilation flags: 

* ASCII, A : Makes several escapes like \w, \b, \s and \d match only on ASCII characters with the respective property.
* DOTALL, S : Make . match any character, including newlines.
* IGNORECASE, I : Do case-insensitive matches.
* LOCALE, L : Do a locale-aware match.
* MULTILINE, M : Multi-line matching, affecting ^ and $.
* VERBOSE, X (for ‘extended’) : Enable verbose REs, which can be organized more cleanly and understandably.

In [174]:
my_string = "Hello World"
pattern = re.compile(r'world', re.IGNORECASE) # No match without I flag
matches = pattern.finditer(my_string)
for match in matches:
    print(match)

my_string = '''
hello
cool
Hello
'''
# line starts with ...
pattern = re.compile(r'^[a-z]', re.MULTILINE) # No match without M flag
matches = pattern.finditer(my_string)
for match in matches:
    print(match)

<re.Match object; span=(6, 11), match='World'>
<re.Match object; span=(1, 2), match='h'>
<re.Match object; span=(7, 8), match='c'>


### End 😔